src/frontend/Deprecation_analysis.ml

```diff
@@ -137,6 +137,40 @@ let rec collect_deprecated_expr (acc : (Location_span.t * string) list)
                 ) ]
           | _ -> [] ) in
       acc @ w @ List.concat_map l ~f:(fun e -> collect_deprecated_expr [] e)
+  | PrefixOp (PNot, ({emeta= {type_= UReal; loc; _}; _} as e)) ->
+      let acc =
+        acc
+        @ [ ( loc
+            , "Using a real as a boolean value is deprecated and will be \
+               disallowed in Stan 2.34. Use an explicit != 0 comparison \
+               instead." ) ] in
+      collect_deprecated_expr acc e
+  | BinOp (({emeta= {type_= UReal; loc; _}; _} as e1), (And | Or), e2)
+   |BinOp (e1, (And | Or), ({emeta= {type_= UReal; loc; _}; _} as e2)) ->
+      let acc =
+        acc
+        @ [ ( loc
+            , "Using a real as a boolean value is deprecated and will be \
+               disallowed in Stan 2.34. Use an explicit != 0 comparison \
+               instead." ) ] in
+      let acc = collect_deprecated_expr acc e1 in
+      let acc = collect_deprecated_expr acc e2 in
+      acc
+  | BinOp ({expr= BinOp (e1, op1, e2); emeta= {loc; _}}, op2, e3)
+    when Operator.(is_cmp op1 && is_cmp op2) ->
+      let pp = Operator.pp in
+      let acc =
+        acc
+        @ [ ( loc
+            , Fmt.str
+                "Found x %a y %a z. This is interpreted as (x %a y) %a z but \
+                 that will change in Stan 2.34. Consider if the intended \
+                 meaning was x %a y && y %a z"
```

Comment:
I don't think the current deprecation policy would allow for this to be changed before 3.0, since it changes the meaning of the same program before and after the version, rather than just being a syntax update.

Reply:
Yes, I wasn't sure if assigning a new meaning would be a good idea. The wording "but that will change in [the future]" is deliberately vague about what the new interpretation will be or if there even is one. In my opinion the current behavior is surprising enough that just making it a syntax error is worth it.

Reply:
You could either say change in 3.0 or become a syntax error on 2.34, but I think we can’t have a program that compiles in 2.21 compile in 2.34 but mean something different